### ISSUE TITLE
Better mapping

### ISSUE THREAD

**Issue by xrotwang** [concepticon/pyconcepticon#3]
The way automatic concept mapping works can be improved, especially from an end-user perspective. @LinguList suggests a number of ways in #398. Amongst those:

- an interactive way of working with concept mapping suggestions (e.g. in the form of a small web app)
- POS information and definitions
- better (more approachable? understandable?) similarity levels
- better handling of malformed input in the `Gloss()` class

**Comment by chrzyki** [comment]
We've also talked about using embedding-based approaches for a new (or different) way of tackling the automated mapping process, see, e.g.:

* https://arxiv.org/pdf/1511.06388.pdf
* http://www.aclweb.org/anthology/P16-1085
* http://aclweb.org/anthology/L18-1162

We've got very valuable training data with all the lists that already have been mapped. Either way, it might be an interesting study/article in itself to do that (i.e. embedding-based mapping) and compare that to our (or other) approaches.

**Comment by chrzyki** [comment]
[This is a simple proof of concept](https://github.com/chrzyki/concepticon-transformers/blob/master/main.py) for learning and predicting glosses using transformers/XLNet. The approach using supersampling/'the whole Concepticon' achieves an F1-score of 0.89 in evaluation, the simple approach using mappings (in `map-en.tsv`) achieves an F1-score of 0.73. Training takes roughly ~60 minutes on a reasonably fast GPU and POS information are used dynamically if available.

@LinguList and I are planning a more thorough evaluation of this, particularly in comparsion to the new approaches in [PySen](https://github.com/lingpy/pysen/blob/master/examples/map.py).

**Comment by LinguList** [comment]
Nice. I am surprised the differences are so high. And I assume: we test on training data, right? So we need to make the cross-validation workflow (which will then be a bit time consuming for the NN, but hopepfully worth the pain) and see how well it works in such a setting.

**Comment by LinguList** [comment]
A, sorry, was not clear: we discuss here the NN trained on all columns for all concept lists vs. the NN trained on map-en.tsv, so we don't even compare with our standard dummy algo as baseline, right?

**Comment by chrzyki** [comment]
Correct. Not yet but it would be good to do that, I think.

The reduced F1 for 'simply' the mappings is a result of the huge number of labels in contrast to not a lot of training data. However, the smaller model is not as overfitted and capable of better generalizations.

**Comment by SimonGreenhill** [comment]
^ this is very cool -- I think you should try write a short paper on this.

**Comment by chrzyki** [comment]
> ^ this is very cool -- I think you should try write a short paper on this.

Thanks! I've received very helpful ideas and notes by @LinguList on how to prepare a comparison of all the different approaches. We'll run those and see how (and where) to go from there.

**Comment by xrotwang** [comment]
@chrzyki assigning this to you, to do whatever you think is appropriate (even if this means closing as "wontfix").

**Comment by LinguList** [comment]
With [pysem](https://github.com/lingpy/pysem), there is now a baseline library, that takes the current version of a concepticon, and offers mappings for this release. So we can probably safely close this for now, or @chrzyki can make a paper out of it, or even share the ideas with Johannes.

**Comment by chrzyki** [comment]
> 
> 
> With [pysem](https://github.com/lingpy/pysem), there is now a baseline library, that takes the current version of a concepticon, and offers mappings for this release. So we can probably safely close this for now, or @chrzyki can make a paper out of it, or even share the ideas with Johannes.

I agree, closing this for now but I'll keep a note of the progress and might use the experience from here (particularly this training file: https://github.com/chrzyki/concepticon-transformers/blob/master/main.py) in combination with `pysem` for a baseline.
